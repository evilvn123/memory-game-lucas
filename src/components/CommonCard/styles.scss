.common-card {
  background-color: transparent;
  perspective: 1000px;
  .inner {
    position: relative;
    width: 100%;
    height: 100%;
    text-align: center;
    transition: transform 0.6s;
    transform-style: preserve-3d;
    &.open {
      transform: rotateY(180deg);
    }
    .common-card-front,
    .common-card-back {
      position: absolute;
      width: 100%;
      height: 100%;
      -webkit-backface-visibility: hidden;
      backface-visibility: hidden;
    }

    .common-card-back {
      transform: rotateY(180deg);
    }
  }
}
