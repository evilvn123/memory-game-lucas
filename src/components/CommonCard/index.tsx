import { CSSProperties, PropsWithChildren, ReactNode } from "react";
import "./styles.scss";

interface Props {
  customStyle?: CSSProperties;
  customClassName?: string;
  open: boolean;
  frontNode?: ReactNode;
  backNode?: ReactNode;
}

export const CommonCard = ({
  customStyle,
  open,
  customClassName,
  frontNode,
  backNode,
}: PropsWithChildren<Props>) => {
  return (
    <div
      data-testid="common-card"
      className={`common-card ${customClassName ? customClassName : ""}`}
      style={customStyle}
    >
      <div data-testid="inner" className={`inner ${open ? "open" : ""}`}>
        <Front>{frontNode}</Front>
        <Back>{backNode}</Back>
      </div>
    </div>
  );
};

export const Front = ({ children }: PropsWithChildren) => {
  return (
    <div data-testid="common-card-front" className="common-card-front">
      {children}
    </div>
  );
};

export const Back = ({ children }: PropsWithChildren) => {
  return (
    <div data-testid="common-card-back" className="common-card-back">
      {children}
    </div>
  );
};
