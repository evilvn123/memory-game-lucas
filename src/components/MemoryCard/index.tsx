import { GameCard } from "../../interfaces/GameCard";
import { CommonCard } from "../CommonCard";
import "./styles.scss";

interface Props {
  card: GameCard;
  onClick: (cardId: string) => void;
}

const MemoryCard = ({ card, onClick }: Props) => {
  return (
    <span
      onClick={() => {
        onClick(card.id);
      }}
      data-testid="memoryCard"
    >
      <CommonCard
        customClassName="memory-card"
        open={card.status === "open" || card.status === "done"}
        frontNode={<div data-testid="front" className="front"></div>}
        backNode={
          <div
            data-testid="back"
            className="back"
            style={{ backgroundImage: `url(${card.image})` }}
          ></div>
        }
      />
    </span>
  );
};

export default MemoryCard;
