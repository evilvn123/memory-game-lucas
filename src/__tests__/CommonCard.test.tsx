import { cleanup, render, screen } from "@testing-library/react";
import { CommonCard, Back, Front } from "../components/CommonCard";
import renderer from "react-test-renderer";

afterEach(cleanup);

it("Render CommonCard with open true", () => {
  render(<CommonCard open={true} />);
  expect(screen.getByTestId("common-card")).toBeInTheDocument();
  expect(screen.getByTestId("inner")).toBeInTheDocument();
  expect(screen.getByTestId("inner")).toHaveClass("open");
});

it("Render CommonCard with open false", () => {
  render(<CommonCard open={false} />);
  expect(screen.getByTestId("common-card")).toBeInTheDocument();
  expect(screen.getByTestId("inner")).toBeInTheDocument();
  expect(screen.getByTestId("inner")).not.toHaveClass("open");
});

it("Render component with customClassName props", () => {
  render(<CommonCard open={false} customClassName="class" />);
  expect(screen.getByTestId("common-card")).toHaveClass("class");
});

it("Render Front face", () => {
  render(<Front />);
  expect(screen.getByTestId("common-card-front")).toBeInTheDocument();
});

it("Render Back face", () => {
  render(<Back />);
  expect(screen.getByTestId("common-card-back")).toBeInTheDocument();
});

it("Render component with customStyle props", () => {
  render(
    <CommonCard
      open={false}
      customStyle={{ background: "rgba(0, 128, 0, 0.3)" }}
    />
  );
  expect(screen.getByTestId("common-card")).toHaveStyle({
    background: "rgba(0, 128, 0, 0.3)",
  });
});

it("renders correctly 1", () => {
  const tree = renderer.create(<CommonCard open={false} />).toJSON();
  expect(tree).toMatchInlineSnapshot(`
<div
  className="common-card "
  data-testid="common-card"
>
  <div
    className="inner "
    data-testid="inner"
  >
    <div
      className="common-card-front"
      data-testid="common-card-front"
    />
    <div
      className="common-card-back"
      data-testid="common-card-back"
    />
  </div>
</div>
`);
  expect(tree).toMatchSnapshot();
});

it("renders correctly 2", () => {
  const tree = renderer.create(<CommonCard open={true} />).toJSON();
  expect(tree).toMatchInlineSnapshot(`
<div
  className="common-card "
  data-testid="common-card"
>
  <div
    className="inner open"
    data-testid="inner"
  >
    <div
      className="common-card-front"
      data-testid="common-card-front"
    />
    <div
      className="common-card-back"
      data-testid="common-card-back"
    />
  </div>
</div>
`);
  expect(tree).toMatchSnapshot();
});

