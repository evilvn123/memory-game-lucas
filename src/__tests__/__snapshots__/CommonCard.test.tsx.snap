// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1 2`] = `
<div
  className="common-card "
  data-testid="common-card"
>
  <div
    className="inner "
    data-testid="inner"
  >
    <div
      className="common-card-front"
      data-testid="common-card-front"
    />
    <div
      className="common-card-back"
      data-testid="common-card-back"
    />
  </div>
</div>
`;

exports[`renders correctly 2 2`] = `
<div
  className="common-card "
  data-testid="common-card"
>
  <div
    className="inner open"
    data-testid="inner"
  >
    <div
      className="common-card-front"
      data-testid="common-card-front"
    />
    <div
      className="common-card-back"
      data-testid="common-card-back"
    />
  </div>
</div>
`;
